File: slp/backend-dotnet/Features/Auth/AuthService.cs
using backend_dotnet.Features.Auth;
using backend_dotnet.Features.Email;
using backend_dotnet.Features.Session;
using backend_dotnet.Features.User;
using System.Security.Claims;

public class AuthService : IAuthService
{
    private readonly IUserRepository _users;
    private readonly ISessionRepository _sessions;
    private readonly IEmailService _email;
    private readonly string _frontendBaseUrl;

    public AuthService(
        IUserRepository users,
        ISessionRepository sessions,
        IEmailService email,
        IConfiguration configuration)
    {
        _users = users;
        _sessions = sessions;
        _email = email;
        _frontendBaseUrl = configuration.GetValue<string>("Frontend:BaseUrlForEmail")?.TrimEnd('/')
                           ?? "http://localhost:3002";
    }

    public async Task<LoginResult> LoginAsync(string username, string password)
    {
        var user = await _users.GetByUsernameAsync(username);
        if (user == null)
        {
            return new LoginResult
            {
                Success = false,
                ErrorCode = "USER_NOT_FOUND",
                Message = "Invalid credentials"
            };
        }

        if (user.Status != "active")
        {
            return new LoginResult
            {
                Success = false,
                ErrorCode = "ACCOUNT_BANNED",
                Message = "Your account has been banned. Please contact support."
            };
        }

        if (!PasswordHasher.Verify(password, user.PasswordHash))
        {
            return new LoginResult
            {
                Success = false,
                ErrorCode = "INVALID_PASSWORD",
                Message = "Invalid credentials"
            };
        }

        var token = SessionTokenService.GenerateToken();
        var tokenHash = SessionTokenService.HashToken(token);

        var session = new Session
        {
            UserId = user.Id,
            TokenHash = tokenHash,
            CreatedAt = DateTime.UtcNow,
            ExpiresAt = DateTime.UtcNow.AddDays(7)
        };

        await _sessions.CreateAsync(session);

        return new LoginResult
        {
            Success = true,
            Data = new LoginResponse
            {
                Token = token,
                UserId = user.Id.ToString(),
                Email = user.Email
            }
        };
    }

    public async Task LogoutAsync(ClaimsPrincipal user)
    {
        var sessionId = user.FindFirst("session_id")?.Value;
        if (sessionId != null)
            await _sessions.RevokeAsync(sessionId);
    }

    public async Task RequestPasswordResetAsync(string email)
    {
        var user = await _users.GetByEmailAsync(email);
        if (user == null) return;

        var token = Guid.NewGuid().ToString();
        user.PasswordResetToken = token;
        user.PasswordResetExpiry = DateTime.UtcNow.AddHours(1);
        await _users.UpdateAsync(user);

        // Build reset link using configured frontend URL
        var resetLink = $"{_frontendBaseUrl}/reset-password?token={token}";
        var htmlBody = EmailTemplates.GetPasswordResetEmail(resetLink);

        await _email.SendHtmlAsync(user.Email, "Reset Your Password", htmlBody);
    }

    public async Task<bool> ConfirmPasswordResetAsync(string token, string newPassword)
    {
        var user = await _users.GetByResetTokenAsync(token);
        if (user == null || user.PasswordResetExpiry < DateTime.UtcNow)
            return false;

        user.PasswordHash = PasswordHasher.Hash(newPassword);
        user.PasswordResetToken = null;
        user.PasswordResetExpiry = null;

        await _users.UpdateAsync(user);

        // Revoke all existing sessions for this user
        await _sessions.RevokeAllForUserAsync(user.Id);

        return true;
    }

    public async Task<bool> VerifyEmailAsync(string token)
    {
        var user = await _users.GetByEmailVerificationTokenAsync(token);
        if (user == null) return false;

        user.EmailConfirmed = true;
        user.EmailVerificationToken = null;

        await _users.UpdateAsync(user);
        return true;
    }

    public async Task SendVerificationEmailAsync(string userId)
    {
        if (!int.TryParse(userId, out var id)) return;

        var user = await _users.GetByIdAsync(id);
        if (user == null) return;

        var token = Guid.NewGuid().ToString();
        user.EmailVerificationToken = token;
        await _users.UpdateAsync(user);

        // Build verification link
        var verifyLink = $"{_frontendBaseUrl}/verify-email?token={token}";
        var htmlBody = EmailTemplates.GetEmailVerificationEmail(verifyLink);

        await _email.SendHtmlAsync(user.Email, "Verify Your Email", htmlBody);
    }

    public async Task<ChangePasswordResult> ChangePasswordAsync(
        string userId,
        string currentPassword,
        string newPassword, string? currentSessionId = null)
    {
        //userId must exist
        if (!int.TryParse(userId, out var id))
        {
            return new ChangePasswordResult
            {
                Success = false,
                ErrorCode = "USER_NOT_FOUND",
                Message = "User not found."
            };
        }

        //User found by userId must also exist
        var user = await _users.GetByIdAsync(id);
        if (user == null)
        {
            return new ChangePasswordResult
            {
                Success = false,
                ErrorCode = "USER_NOT_FOUND",
                Message = "User not found."
            };
        }

        //Current password user input must exist
        if (string.IsNullOrEmpty(currentPassword))
        {
            return new ChangePasswordResult
            {
                Success = false,
                ErrorCode = "INVALID_CURRENT_PASSWORD",
                Message = "Current password is required."
            };
        }

        //Current password in database must exist
        if (string.IsNullOrEmpty(user.PasswordHash))
        {
            return new ChangePasswordResult
            {
                Success = false,
                ErrorCode = "NO_LOCAL_PASSWORD",
                Message = "User does not have a password set."
            };
        }

        //New password must exist
        if (string.IsNullOrEmpty(newPassword))
        {
            return new ChangePasswordResult
            {
                Success = false,
                ErrorCode = "NO_NEW_PASSWORD",
                Message = "New password null."
            };
        }

        //Current password of user input vs database must match
        if (!PasswordHasher.Verify(currentPassword, user.PasswordHash))
        {
            return new ChangePasswordResult
            {
                Success = false,
                ErrorCode = "INVALID_CURRENT_PASSWORD",
                Message = "Current password is incorrect."
            };
        }

        //New password must not match current password
        if (currentPassword == newPassword)
        {
            return new ChangePasswordResult
            {
                Success = false,
                ErrorCode = "INVALID_CURRENT_PASSWORD",
                Message = "Current password and new password must different."
            };
        }

        user.PasswordHash = PasswordHasher.Hash(newPassword);
        user.UpdatedAt = DateTime.UtcNow;
        await _users.UpdateAsync(user);

        // Revoke all OTHER sessions; keep current one so user stays logged in
        if (currentSessionId != null)
            await _sessions.RevokeAllForUserExceptAsync(user.Id, currentSessionId);
        else
            await _sessions.RevokeAllForUserAsync(user.Id);

        return new ChangePasswordResult { Success = true };
    }
}

File: slp/backend-java/src/main/java/com/app/backendjava/features/auth/AuthService.java
package com.app.backendjava.features.auth;

import com.app.backendjava.features.user.*;
import com.app.backendjava.features.session.*;
import com.app.backendjava.features.email.*;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.security.core.Authentication;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.UUID;

@Service
@RequiredArgsConstructor
public class AuthService implements IAuthService {
    private final IUserRepository userRepository;
    private final SessionRepository sessionRepository;
    private final EmailService emailService;

    @Value("${frontend.base-url:http://localhost:3002}")
    private String frontendBaseUrl;

    @Override
    @Transactional
    public AuthDto.LoginResult loginAsync(String username, String password) {
        var user = userRepository.getByUsernameAsync(username).orElse(null);

        if (user == null) {
            return AuthDto.LoginResult.builder()
                    .success(false).errorCode("USER_NOT_FOUND").message("Invalid credentials").build();
        }

        if (!"active".equals(user.getStatus())) {
            return AuthDto.LoginResult.builder()
                    .success(false).errorCode("ACCOUNT_BANNED")
                    .message("Your account has been banned. Please contact support.").build();
        }

        if (!PasswordHasher.verify(password, user.getPasswordHash())) {
            return AuthDto.LoginResult.builder()
                    .success(false).errorCode("INVALID_PASSWORD").message("Invalid credentials").build();
        }

        String rawToken = SessionTokenService.generateToken();
        String hashedToken = SessionTokenService.hashToken(rawToken);

        Session session = Session.builder()
                .id(UUID.randomUUID().toString())
                .userId(user.getId())
                .tokenHash(hashedToken)
                .createdAt(LocalDateTime.now())
                .expiresAt(LocalDateTime.now().plusDays(7))
                .revoked(false)
                .build();

        sessionRepository.save(session);

        AuthDto.LoginResponse response = new AuthDto.LoginResponse();
        response.setToken(rawToken); // Send the raw token to the client
        response.setUserId(user.getId().toString());
        response.setEmail(user.getEmail());

        return AuthDto.LoginResult.builder().success(true).data(response).build();
    }

    @Override
    @Transactional
    public void logoutAsync(Authentication auth) {
        if (auth != null && auth.getDetails() instanceof String sessionId) {
            sessionRepository.revokeAsync(sessionId);
        }
    }

    @Override
    @Transactional
    public void requestPasswordResetAsync(String email) {
        userRepository.getByEmailAsync(email).ifPresent(user -> {
            String token = UUID.randomUUID().toString();
            user.setPasswordResetToken(token);
            user.setPasswordResetExpiry(LocalDateTime.now().plusHours(1));
            userRepository.saveAsync(user);

            String resetLink = frontendBaseUrl + "/reset-password?token=" + token;
            emailService.sendHtmlAsync(user.getEmail(), "Reset Your Password", resetLink);
        });
    }

    @Override
    @Transactional
    public boolean confirmPasswordResetAsync(String token, String newPassword) {
        var user = userRepository.getByResetTokenAsync(token).orElse(null);
        if (user == null || user.getPasswordResetExpiry().isBefore(LocalDateTime.now())) {
            return false;
        }

        user.setPasswordHash(PasswordHasher.hash(newPassword));
        user.setPasswordResetToken(null);
        user.setPasswordResetExpiry(null);
        userRepository.saveAsync(user);

        sessionRepository.revokeAllForUserAsync(user.getId());
        return true;
    }

    @Override
    @Transactional
    public boolean verifyEmailAsync(String token) {
        var user = userRepository.getByEmailVerificationTokenAsync(token).orElse(null);
        if (user == null) return false;

        user.setEmailConfirmed(true);
        user.setEmailVerificationToken(null);
        userRepository.saveAsync(user);
        return true;
    }

    @Override
    @Transactional
    public void sendVerificationEmailAsync(String userId) {
        userRepository.getByIdAsync(Integer.parseInt(userId)).ifPresent(user -> {
            String token = UUID.randomUUID().toString();
            user.setEmailVerificationToken(token);
            userRepository.saveAsync(user);

            String verifyLink = frontendBaseUrl + "/verify-email?token=" + token;
            emailService.sendHtmlAsync(user.getEmail(), "Verify Your Email", verifyLink);
        });
    }

    @Override
    @Transactional
    public AuthDto.ChangePasswordResult changePasswordAsync(String userId, String currentPassword, String newPassword, String currentSessionId) {
        var user = userRepository.getByIdAsync(Integer.parseInt(userId)).orElse(null);
        if (user == null) return AuthDto.ChangePasswordResult.builder().success(false).errorCode("USER_NOT_FOUND").message("User not found.").build();

        if (!PasswordHasher.verify(currentPassword, user.getPasswordHash())) {
            return AuthDto.ChangePasswordResult.builder().success(false).errorCode("INVALID_CURRENT_PASSWORD").message("Current password incorrect.").build();
        }

        user.setPasswordHash(PasswordHasher.hash(newPassword));
        user.setUpdatedAt(LocalDateTime.now());
        userRepository.saveAsync(user);

        if (currentSessionId != null) {
            sessionRepository.revokeAllForUserExceptAsync(user.getId(), currentSessionId);
        } else {
            sessionRepository.revokeAllForUserAsync(user.getId());
        }

        return AuthDto.ChangePasswordResult.builder().success(true).build();
    }
}